qiskit/extensions/quantum_initialiser/initialiser.py

```diff
@@ -0,0 +1,411 @@
+            return bottom_gate(list_of_angles[0], target_qubit)
+
+        local_num_qubits = Initialiser._log2(list_len) + 1
+        control_qubit = self.nth_qubit_from_least_sig_qubit(local_num_qubits - 1 + bottom_qubit_index)
+
+        # already checked:
+        # if local_num_qubits == 1:
+        #     raise QISKitException("Number of angles not a power of 2 :", list_len)
+
+        # Case of one or more rounds of multiplexing
+        # if local_num_qubits > 1:
+
+        # calc angle weights, assuming recursion (that is the lower-level requested angles have been correctly
+        # implemented by recursion)
+        angle_weight = scipy.kron([[0.5, 0.5], [0.5, -0.5]], numpy.identity(2 ** (local_num_qubits - 2)))
+
+        # calc the combo angles
+        list_of_angles = (angle_weight * numpy.matrix(list_of_angles).transpose()).reshape(-1).tolist()[0]
+
+        combine_composite_gates = CompositeGate("multiplex" + local_num_qubits.__str__(), [], self.arg)
+
+        # recursive step on half the angles fulfilling the above assumption
+        combine_composite_gates._attach(
+            self._multiplex(bottom_gate, bottom_qubit_index, list_of_angles[0:(list_len // 2)]))
+
+        # combine_composite_gates.cx(control_qubit,target_qubit) -> does not work as expected because checks circuit
+        # so attach CNOT as follows, thereby flipping the LSB qubit
+        combine_composite_gates._attach(CnotGate(control_qubit, target_qubit))
+
+        # implement extra efficiency from the paper of cancelling adjacent CNOTs (by leaving out last CNOT and
+        # reversing (NOT inverting) the second lower-level multiplex)
+        sub_gate = self._multiplex(bottom_gate, bottom_qubit_index, list_of_angles[(list_len // 2):])
+        if isinstance(sub_gate, CompositeGate):
+            combine_composite_gates._attach(sub_gate.reverse())
+        else:
+
+            combine_composite_gates._attach(sub_gate)
+
+        # outer multiplex keeps final CNOT, because no adjacent CNOT to cancel with
+        if self.num_qubits == local_num_qubits + bottom_qubit_index:
+            combine_composite_gates._attach(CnotGate(control_qubit, target_qubit))
+
+        # commenting out the non-optimised way of not cancelling out CNOTs:
+        # combine_composite_gates._attach(
+        #   self._multiplex(bottom_gate, bottom_qubit_index, list_of_angles[(list_len // 2):]))
+        # combine_composite_gates._attach(CnotGate(control_qubit, target_qubit))
+
+        return combine_composite_gates
+
+    @staticmethod
+    def chop_num(numb):
+        """Set very small numbers (as defined by global variable EPS) to zero"""
+        return 0 if abs(numb) < EPS else numb
+
+
+# ###############################################################
+# Add needed functionality to other classes (it feels
+#  weird following the QISKit convention of adding functionality to other classes like this ;),
+#  TODO: multiple inheritance might be better?)
```

Comment:
> it feels weird following the QISKit convention of adding functionality to other classes like this

It feels a bit unusual indeed, and comes at a discoverability/readability/maintainability cost that you seem to have experienced already, specially due to the high number of functions in the PRs - thanks for raising the issue! It seems a bit out of the scope of this PR to revise the convention, but I'll make a note so we can revise it eventally,